datafusion/common/src/tree_node.rs

```diff
@@ -353,13 +425,38 @@ pub trait TreeNode: Sized {
     }
 
     /// Apply the closure `F` to the node's children.
+    ///
+    /// See `mutate_children` for rewriting in place
     fn apply_children<F: FnMut(&Self) -> Result<TreeNodeRecursion>>(
         &self,
         f: &mut F,
     ) -> Result<TreeNodeRecursion>;
 
-    /// Apply transform `F` to the node's children. Note that the transform `F`
-    /// might have a direction (pre-order or post-order).
+    /// Rewrite the node's children in place using `F`.
```

Comment:
I chose to leave `mutate_children` unimplemented for `Expr` and the other `TreeNodes` initially to reduce the size of this PR (and thus the review burden). I will file a follow on ticket to track doing so